### PR TITLE
Add `shm://` rzpipe test

### PR DESCRIPTION
**Pull request by XVilka**
# DO NOT SQUASH

**Your checklist for this pull request**
- [x] I've read the [guidelines for contributing](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md) to this repository
- [x] I made sure to follow the project's [coding style](https://github.com/rizinorg/rizin/blob/master/DEVELOPERS.md#code-style)
- [x] I've added tests that prove my fix is effective or that my feature works (if possible)
- [ ] I've updated the documentation and the [rizin book](https://github.com/rizinorg/book) with the relevant information (if needed)

**Detailed description**

Continuation of https://github.com/rizinorg/rizin/pull/1759

**Test plan**

CI is green

### CODE REVIEW

**Comment by XVilka** [comment]
For old Debians and CentOS need to install this, and modify the script to use the backport of `shared_memory` to Python versions earlier than 3.8 https://github.com/dillonalaird/shared_numpy

**Comment by XVilka** [comment]
@devnexen maybe you could take a look if it works also on DragonFlyBSD and Haiku. Also, do you have any clue why this test works on NetBSD but fails with FreeBSD and OpenBSD?
```
[XX] db/cmd/cmd_pipe rzpipe.py - shared memory (shm://)
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc '?q `env~?^ASAN=1$`
?+ env LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libasan.so.5
#!pipe python3 scripts/shared_memory.py
' 'bins/elf/_Exit (42)'
-- stdout
--- expected
+++ actual
@@ -3,10 +3,3 @@
 -------------
 Shared buffer size 0x360c
 -------------
-{'arch': 'x86', 'baddr': 134512640, 'binsz': 13073, 'bintype': 'elf', 'bits': 32, 'class': 'ELF32', 'compiler': 'GCC: (GNU) 9.2.1 20190827 (Red Hat 9.2.1-1)', 'endian': 'LE', 'intrp': '/lib/ld-linux.so.2', 'laddr': 0, 'lang': 'c++', 'machine': 'Intel 80386', 'maxopsz': 16, 'minopsz': 1, 'os': 'linux', 'pcalign': 0, 'relro': 'partial', 'rpath': 'NONE', 'subsys': 'linux', 'stripped': False, 'crypto': False, 'havecode': True, 'va': True, 'sanitiz': False, 'static': False, 'linenum': True, 'lsyms': True, 'canary': False, 'PIE': False, 'RELROCS': True, 'NX': True}
-- offset -   0 1  2 3  4 5  6 7  8 9  A B  C D  E F  0123456789ABCDEF
-0x08049020  5589 e583 ec08 83ec 0c6a 2ae8 e0ff ffff  U........j*.....
-
-0x08049020    1 16           entry0
-0x08049010    1 6            sym.imp._Exit
-

-- stderr
WARNING: Neither hash nor gnu_hash exist. Falling back to heuristics for deducing the number of dynamic symbols...
WARNING: Neither hash nor gnu_hash exist. Falling back to heuristics for deducing the number of dynamic symbols...
WARNING: Neither hash nor gnu_hash exist. Falling back to heuristics for deducing the number of dynamic symbols...
ERROR: Cannot connect to shared memory "psm_0917d541" (0x12b370ef)
[r] Cannot open 'shm://psm_0917d541'
Traceback (most recent call last):
  File "scripts/shared_memory.py", line 19, in <module>
    rzp.cmd("e scr.color=0")
  File "/usr/local/lib/python3.8/site-packages/rzpipe/open_base.py", line 244, in cmd
    res = self._cmd(cmd, **kwargs)
  File "/usr/local/lib/python3.8/site-packages/rzpipe/open_sync.py", line 104, in _cmd_process
    self.process.stdin.write((cmd + "\n").encode("utf8"))
BrokenPipeError: [Errno 32] Broken pipe
/usr/local/lib/python3.8/multiprocessing/resource_tracker.py:216: UserWarning: resource_tracker: There appear to be 1 leaked shared_memory objects to clean up at shutdown
  warnings.warn('resource_tracker: There appear to be %d '
```

@thestr4ng3r could you please take a look for this on MacOS?
```
Traceback (most recent call last):
  File "/Users/runner/work/rizin/rizin/test/scripts/shared_memory.py", line 12, in <module>
    shm.buf[:] = data[:]
ValueError: memoryview assignment: lvalue and rvalue have different structures
```

**Comment by devnexen** [comment]
is `shm_open` having correct naming (starting with slash) ?

**Comment by XVilka** [comment]
@devnexen that might be the case, it doesn't start with slash.

**Comment by XVilka** [comment]
Fixed the problem for all systems except MacOS, OpenBSD, and Windows.

**Comment by devnexen** [comment]
Oddily I could not reproduce locally on openbsd the same failure (yes I set malloc_options too).

File: librz/io/p/io_shm.c [hunk 1]
```diff
@@ -106,10 +108,10 @@ static RzIODesc *shm__open(RzIO *io, const char *pathname, int rw, int mode) {
 		return NULL;
 	}
 	const char *ptr = pathname + 6;
-	shm->name = strdup(ptr);
+	shm->name = rz_str_newf("/%s", ptr);
```
Comment (ret2libc) [inline librz/io/p/io_shm.c]: why this?

Comment (devnexen) [reply librz/io/p/io_shm.c]: for some oses it s mandatory the name starts with a slash to work correctly.

Comment (ret2libc) [reply librz/io/p/io_shm.c]: isn't that accomplished by just doing `shm:///dev/shm/hello`?

Comment (XVilka) [reply librz/io/p/io_shm.c]: @ret2libc I'ts unportable, I think. Anyway, this method works everywhere except MacOS and Windows (https://github.com/rizinorg/rizin/issues/1758)
I have zero idea why OpenBSD CI fails though, maybe not enough memory to do the `mmap()`?

Comment (devnexen) [reply librz/io/p/io_shm.c]: that might be it however when I get OOM issue on OpenBSD with software X I usually get it on NetBSD too.

Comment (ret2libc) [reply librz/io/p/io_shm.c]: How does the `shm://<xxxxx>` string look like for all major platforms?

Like, for Linux it is something like `shm:///dev/shm/hello`... For the others?

Comment (XVilka) [reply librz/io/p/io_shm.c]: It doesn't look like `shm:///dev/shm/hello` - it's `shm://psm_7effab` on Linux, for example

Comment (ret2libc) [reply librz/io/p/io_shm.c]: Ok, the others?

**Comment by thestr4ng3r** [comment]
On macOS, the size is always rounded up to be page-aligned it seems. Maybe we have to pass the size manually somehow then.

**Comment by XVilka** [comment]
Regarding the size, I wonder if we can code the buffer size in the URI itself, something like `shm://psm_7effab:0x345` where `0x345` is the buffer size.

@thestr4ng3r curiously, there is a difference in handling PE files, could you please take a look when you have time?:
```
[XX] db/archos/not-windows-any/cmd_pipe rzpipe.py - shared memory (shm://)
RZ_NOPLUGINS=1 rizin -escr.utf8=0 -escr.color=0 -escr.interactive=0 -N -Qc '?q `env~?^ASAN=1$`
?+ env LD_PRELOAD=/usr/lib/x86_64-linux-gnu/libasan.so.5
#!pipe python3 scripts/shared_memory.py
' 'bins/elf/_Exit (42)'
-- stdout
--- expected
+++ actual
@@ -1,7 +1,7 @@
 Copying bins/elf/_Exit (42)...
 Copied bins/elf/_Exit (42) succesfully
 -------------
-Shared buffer size 0x360c
+Shared buffer size 0x4000
 -------------
 {'arch': 'x86', 'baddr': 134512640, 'binsz': 13073, 'bintype': 'elf', 'bits': 32, 'class': 'ELF32', 'compiler': 'GCC: (GNU) 9.2.1 20190827 (Red Hat 9.2.1-1)', 'endian': 'LE', 'intrp': '/lib/ld-linux.so.2', 'laddr': 0, 'lang': 'c++', 'machine': 'Intel 80386', 'maxopsz': 16, 'minopsz': 1, 'os': 'linux', 'pcalign': 0, 'relro': 'partial', 'rpath': 'NONE', 'subsys': 'linux', 'stripped': False, 'crypto': False, 'havecode': True, 'va': True, 'sanitiz': False, 'static': False, 'linenum': True, 'lsyms': True, 'canary': False, 'PIE': False, 'RELROCS': True, 'NX': True}
 - offset -   0 1  2 3  4 5  6 7  8 9  A B  C D  E F  0123456789ABCDEF
@@ -24,37 +24,33 @@
 Copying bins/pe/winver.exe...
 Copied bins/pe/winver.exe succesfully
 -------------
-Shared buffer size 0x400
+Shared buffer size 0x1000
 -------------
 {'arch': 'x86', 'baddr': 4194304, 'binsz': 4294967295, 'bintype': 'pe', 'bits': 32, 'retguard': False, 'class': 'PE32', 'cmp.csum': '0x00000000', 'compiled': 'Thu Jan  1 00:00:00 1970 UTC', 'endian': 'LE', 'hdr.csum': '0x00000000', 'laddr': 0, 'lang': 'c', 'machine': 'i386', 'maxopsz': 16, 'minopsz': 1, 'os': 'windows', 'overlay': True, 'cc': 'cdecl', 'pcalign': 0, 'signed': False, 'subsys': 'Windows CUI', 'stripped': False, 'crypto': False, 'havecode': True, 'va': True, 'sanitiz': False, 'static': False, 'linenum': False, 'lsyms': False, 'canary': False, 'PIE': False, 'RELROCS': False, 'NX': False}
 - offset -   0 1  2 3  4 5  6 7  8 9  A B  C D  E F  0123456789ABCDEF
 0x00401000  6880 1040 00ff 15b4 1140 00ff 3590 1040  h..@.....@..5..@
 
-0x00401000    1 59           entry0
-0x004011b4    1 8            fcn.004011b4
-0x004011bc    8 3653 -> 7171 fcn.004011bc
-0x004011b0    1 17           fcn.004011b0
+0x00401000    1 58           entry0
 
             ;-- section.sect_0:
 / entry0 ();
-|           0x00401000      push  0x401080                             ; [00] -rwx section size 4096 named sect_0
-|           0x00401005      call  dword [fcn.004011b4]                 ; 0x4011b4
+|           0x00401000      push  0x401080                             ; LPOSVERSIONINFOA lpVersionInformation ; [00] -rwx section size 4096 named sect_0
+|           0x00401005      call  dword [sym.imp.kernel32.dll_GetVersionExA] ; 0x4011b4 ; BOOL GetVersionExA(LPOSVERSIONINFOA lpVersionInformation)
 |           0x0040100b      push  dword [0x401090]
 |           0x00401011      push  dword [0x40108c]
 |           0x00401017      push  dword [0x401088]
 |           0x0040101d      push  dword [0x401084]
-|           0x00401023      push  0x40103c                             ; " * a PE overriding OS values: OS Ver %i.%i.%i PlatformID %i\n\n" ; int32_t arg_73h
-|           0x00401028      call  dword [fcn.004011bc]                 ; 0x4011bc
+|           0x00401023      push  0x40103c                             ; " * a PE overriding OS values: OS Ver %i.%i.%i PlatformID %i\n\n" ; const char *format
+|           0x00401028      call  dword [sym.imp.msvcrt.dll_printf]    ; 0x4011bc ; int printf(const char *format)
 |           0x0040102e      add   esp, 0x10
 |           0x00401031      nop
-|           0x00401032      push  0
-|           0x00401034      call  dword [fcn.004011b0]                 ; 0x4011b0
-\           0x0040103a      int3
+|           0x00401032      push  0                                    ; UINT uExitCode
+\           0x00401034      call  dword [sym.imp.kernel32.dll_ExitProcess] ; 0x4011b0 ; VOID ExitProcess(UINT uExitCode)
 
 Copying bins/mach0/mach0_2-x86_64...
 Copied bins/mach0/mach0_2-x86_64 succesfully
 -------------
-Shared buffer size 0x10e8
+Shared buffer size 0x2000
 -------------
 {'arch': 'x86', 'baddr': 4294967296, 'binsz': 4294967295, 'bintype': 'mach0', 'bits': 64, 'class': 'MACH064', 'compiler': '', 'endian': 'LE', 'intrp': '/usr/lib/dyld', 'laddr': 0, 'lang': 'c', 'machine': 'x86 64 all', 'maxopsz': 16, 'minopsz': 1, 'os': 'macos', 'pcalign': 0, 'subsys': 'darwin', 'stripped': False, 'crypto': False, 'havecode': True, 'va': True, 'sanitiz': False, 'static': False, 'linenum': False, 'lsyms': False, 'canary': False, 'PIE': True, 'RELROCS': False, 'NX': False}
 - offset -    0 1  2 3  4 5  6 7  8 9  A B  C D  E F  0123456789ABCDEF

-- stderr
```

**Comment by thestr4ng3r** [comment]
> Regarding the size, I wonder if we can code the buffer size in the URI itself, something like `shm://psm_7effab:0x345` where `0x345` is the buffer size.

I was thinking this too, either `shm://psm_7effab:0x345` or `shm://psm_7effab/0x345` (maybe more consistent with other plugins)

**Comment by thestr4ng3r** [comment]
There are differences between regular open and shm even on linux:

Bildschirmfoto 2021-10-11 um 20 19 50

**Comment by XVilka** [comment]
@thestr4ng3r @ret2libc @wargio please take a look again, it passes all CIs now and is ready and consistent everywhere. I would like to cherry-pick this for 0.3.1 as well.

**Comment by XVilka** [comment]
Windows installer error is unrelated cc @ret2libc:
```
Parsing [Files] section, line 61
Error on line 61 in D:\a\rizin\rizin\dist\windows\rizin.iss: No files found matching "C:\Users\RUNNER~1\AppData\Local\Temp\rizin-win-installer\*"
Compile aborted.
Remove-Item: D:\a\rizin\rizin\dist\windows\build_windows_installer.ps1:27
Line |
  27 |  rm -r $installdir
     |  ~~~~~~~~~~~~~~~~~
     | Cannot find path 'C:\Users\RUNNER~1\AppData\Local\Temp\rizin-win-installer' because it does not exist.

Error: Process completed with exit code 1.
```

**Review by ret2libc — APPROVED** [review]

**Review by thestr4ng3r — APPROVED** [review]

**Pull request merged**